src/main/scala/viper/silver/plugin/standard/reasoning/ReasoningPlugin.scala

```diff
@@ -0,0 +1,313 @@
+        val new_trigs = trigs.map(t => Trigger(t.exps.map(e1 => applySubstitution(new_v_map, e1)))(t.pos))
+        Seqn(
+          Seq(
+            Assert(Exists(new_v_map.map(_._2), new_trigs, new_exp)(e.pos, ReasoningInfo))(e.pos)
+          )
+            ++
+            v.map(variable => LocalVarDeclStmt(variable)(variable.pos)) //list of variables
+            ++
+            Seq(
+              Inhale(exp)(e.pos)
+            ),
+          Seq()
+        )(e.pos)
+
+      case u@UniversalIntro(v, trigs, exp1, exp2, blk) =>
+        val boolvar = LocalVarDecl(uniqueName("b", usedNames), Bool)(exp1.pos)
+
+        /** Get all variables that are in scope in the current method */
+        val tainted: Set[LocalVarDecl] = v.toSet
+        val varsOutside = (input.methods
+          .filter(m => m.body.isDefined)
+          .flatMap(m => m.body.get.ss.filter(s => s.contains(u)).flatMap(_ => Set(m.transitiveScopedDecls: _*))).toSet -- Set(u.transitiveScopedDecls: _*)) ++ tainted
+        /**
+          * get all variables that are assigned to inside the block and take intersection with universal introduction
+          * variables. If they are contained throw error since quantified variables should be immutable
+          */
+        val writtenVars: Set[LocalVarDecl] = analysis.getModifiedVars(blk).collect({ case v: LocalVarDecl => v})
+        checkReassigned(writtenVars, v, reportError, u)
+        checkInfluencedBy(input, reportError)
+
+        /** Contains all variables that must not be tainted */
+        val volatileVars: Set[LocalVarDecl] = analysis.getLocalVarDeclsFromExpr(exp1) ++ analysis.getLocalVarDeclsFromExpr(exp2) -- v
+        /** execute modular flow analysis using graph maps for the universal introduction statement */
+        analysis.executeTaintedGraphAnalysis(varsOutside.collect({ case v:LocalVarDecl => v }), tainted, blk, volatileVars, u)
+
+        /** Translate the new syntax into Viper language */
+        val (newVarMap, newExp1) = substituteWithFreshVars(v, exp1, usedNames)
+        val newExp2 = applySubstitution(newVarMap, exp2)
+        val quantifiedVars = newVarMap.map(vars => vars._2)
+        val newTrigs = trigs.map(t => Trigger(t.exps.map(e1 => applySubstitution(newVarMap, e1)))(t.pos))
+        val lbl = uniqueName("l", usedNames)
+
+        Seqn(
+          Seq(
+            Label(lbl, Seq())(u.pos),
+            If(boolvar.localVar,
+              Seqn(
+                Seq(
+                  Inhale(exp1)(exp1.pos)
+                ),
+                Seq())(exp1.pos),
+              Seqn(Seq(), Seq())(exp1.pos)
+
+            )(exp1.pos),
+            blk,
+            Assert(Implies(boolvar.localVar, exp2)(exp2.pos))(exp2.pos),
+            Inhale(Forall(quantifiedVars, newTrigs, Implies(LabelledOld(newExp1, lbl)(exp2.pos), newExp2)(exp2.pos))(exp2.pos))(exp2.pos)
+          ),
+          Seq(boolvar) ++ v
+        )(exp1.pos)
+
```

Comment:
```suggestion
        val errTransformer = ErrTrafo({
          case AssertFailed(_, r, c) => UniversalIntroFailed(u, r, c)
          case d => d
        })

        Seqn(
          Seq(
            Label(lbl, Seq.empty)(u.pos),
            // conditionally inhale assume expression
            If(boolvar.localVar,
              Seqn(
                Seq(Inhale(exp1)(exp1.pos)),
                Seq.empty)(exp1.pos),
              Seqn(Seq.empty, Seq.empty)(exp1.pos)
            )(exp1.pos),
            // execute block
            blk,
            // conditionally assert imply expression
            Assert(Implies(boolvar.localVar, exp2)(exp2.pos))(exp2.pos, errT = errTransformer),
            Inhale(Forall(quantifiedVars, newTrigs, Implies(LabelledOld(newExp1, lbl)(exp2.pos), newExp2)(exp2.pos))(exp2.pos))(exp2.pos)
          ),
          Seq(boolvar) ++ v
        )(exp1.pos)
```

Reply:
This snippet still requires some field renaming due to the suggested changes for `UniversalIntro` earlier